facility/list-facility-structure.js

```diff
@@ -0,0 +1,184 @@
+
+    // STEP 3 - collect assets and related room references
+    for (const link of facility.links) {
+        // skip default model
+        if (isDefaultModel(FACILITY_URN, link.modelId)) {
+            continue;
+        }
+        const assets = await client.getTaggedAssets(link.modelId,
+            [ ColumnFamilies.Standard, ColumnFamilies.DtProperties, ColumnFamilies.Refs, ColumnFamilies.Xrefs ]);
+        const modelId = link.modelId;
+
+        for (const asset of assets) {
+            // unique asset key
+            const assetKey = Encoding.toFullKey(asset[QC.Key]);
+
+            data.assets[assetKey] = asset;
+            // STEP 4 - find room references. Note that reference can be within same model or across models
+            let roomRef = asset[QC.Rooms];
+            const assetRooms = [];
+
+            if (roomRef) {
+                const roomKeys = Encoding.fromShortKeyArray(roomRef);
+
+                for (const roomKey of roomKeys) {
+                    assetRooms.push({
+                        modelId: modelId,
+                        roomId: Encoding.toFullKey(roomKey)
+                    });
+                }
+            } else {
+                roomRef = asset[QC.XRooms];
+                const roomKeys = Encoding.fromXrefKeyArray(roomRef);
+                const modelIds = roomKeys[0];
+                const elementKeys = roomKeys[1];
+
+                for (let i = 0; i < modelIds.length; i++) {
+                    assetRooms.push({
+                        modelId: `urn:adsk.dtm:${modelIds[i]}`,
+                        roomId: elementKeys[i]
+                    });
+                }
+            }
+            // STEP 5 - build map between asset and rooms - note that asset can be linked to more than one room
+            for (const { roomId } of assetRooms) {
+                const roomKey = roomId;
+                let assetIds = data.roomAssetsMap[roomKey];
+
+                if (!assetIds) {
+                    assetIds = [];
+                }
+                assetIds.push(assetKey);
+                data.roomAssetsMap[roomKey] = assetIds;
+            }
+            modelRooms.push(... assetRooms);
+        }
+    }
+    // STEP 6 - process rooms and create map between room and level
+    const modelIds = new Set(modelRooms.map(i => i.modelId));
+
+    for (const modelId of modelIds) {
+        const roomIds = new Set(modelRooms.filter(i => i.modelId === modelId).map(i => i.roomId));
+        const rooms = await client.getElements(modelId, [... roomIds ],
+            [ ColumnFamilies.Standard, ColumnFamilies.Refs ]);
+        const levelIds = new Set();
+
+        for (const room of rooms) {
+            const roomKey = Encoding.toFullKey(room[QC.Key]);
+
+            data.rooms[roomKey] = room;
+            const levelRef = room[QC.Level];
+
+            if (levelRef) {
+                levelIds.add(levelRef);
+                data.roomLevelMap[roomKey] = Encoding.toFullKey(levelRef);
```

Comment:
same here, level is logical. Though I do not really understand why do we need to re-encode it.